lib/datasource/index.ts

```diff
@@ -55,25 +59,139 @@ const datasources: Record<string, Datasource> = {
+    url = url.replace('www.github.com', 'github.com');
+    // istanbul ignore if
+    if (url.startsWith('http://github.com/')) {
+      url = 'https://' + url.substr(7);
+    }
+    //  istanbul ignore if
+    if (url.startsWith('https://github.com/')) {
+      url = url
+        .split('/')
+        .slice(0, 5)
+        .join('/');
+    } // a lot of this is probably redundant and can be better achieved with URL
+    if (
+      getHostsFromRulesGithub === undefined ||
+      !getHostsFromRulesGithub.includes(URL.parse(url).hostname)
+    ) {
+      extraBaseUrls.push('github.com', 'gist.github.com');
+    } else {
+      getHostsFromRulesGithub.forEach(host => {
+        extraBaseUrls.push(host, `gist.${host}`);
+      });
+    }
+    url = parse(url, {
+      extraBaseUrls,
+    });
+    if (url !== null || url !== undefined) {
+      return url;
+    }
+  }
+  if (parsedUrl.protocol && parsedUrl.host && parsedUrl.path) {
+    let tmpProtocol: string;
+    if (parsedUrl.protocol !== 'http:') {
+      tmpProtocol = 'https:';
+    } else {
+      tmpProtocol = 'http:';
+    }
+    const getHostsFromRulesGitlab =
+      hostRules.hosts({ hostType: 'gitlab' }) || [];
+    const getHostsFromRulesBitbucket =
+      hostRules.hosts({ hostType: 'bitbucket' }) || [];
+    const getHostsFromRulesBitbucketServer =
+      hostRules.hosts({ hostType: 'bitbucket-server' }) || [];
+    const gitInfo = hostedGitInfo.fromUrl(url) || null;
+    let tmpurl: string;
+    if (gitInfo) {
+      tmpurl = gitInfo.browse({ noGitPlus: true, noCommittish: true });
+    }
+    if (
+      getHostsFromRulesGitlab.includes(parsedUrl.host) ||
+      parsedUrl.host === 'gitlab.com'
+    ) {
+      if (gitInfo && gitInfo.type === 'gitlab') {
+        url = tmpurl;
+      } else {
+        url = `${tmpProtocol}//${parsedUrl.host}${parsedUrl.path
+          .replace(RegExp('/api/v[3|4]/'), '')
+          .replace(':', '')
+          .replace(RegExp('.git$'), '')
+          .split('/')
+          .join('/')
+          .replace(RegExp('/$'), '')}`;
+      }
+    } else if (
+      getHostsFromRulesBitbucket.includes(parsedUrl.host) ||
+      parsedUrl.host === 'bitbucket.org' ||
+      getHostsFromRulesBitbucketServer.includes(parsedUrl.host)
+    ) {
+      if (gitInfo && gitInfo.type === 'bitbucket') {
+        url = tmpurl;
+      } else {
+        url = `${tmpProtocol}//${parsedUrl.host}${parsedUrl.path
+          .replace(':', '')
+          .split('/')
+          .slice(0, 3)
+          .join('/')
+          .replace(RegExp('.git$'), '')
+          .replace(RegExp('/$'), '')}`;
+      }
+    }
+    return url.replace(RegExp('.git$'), '').replace(RegExp('/$'), '');
+  }
+  /* todo add azure support. */
+  return null;
+}
+export async function fetchReleases(
   config: PkgReleaseConfig
 ): Promise<ReleaseResult | null> {
   const { datasource } = config;
+  // istanbul ignore if
   if (!datasource) {
     logger.warn('No datasource found');
-    return null;
```

Comment:
revert this, otherwise we can get exception in next `if`. We also should have a test for it, so ignore should not required.